Adds Packet conversion semantics.

micolous opened the pull request:
This pull requests adds Packet conversion semantics:

* `Packet.convert_to` (method) allows a Packet to declare how it can be   converted into another Packet type.
* `Packet.convert_packet` (classmethod) allows a Packet to declare how it convert other Packet instances into itself.
* Both methods will fall back to ~getting child layers, or~ using the other method.  In the absence of available conversions, it ~returns `None`~ raises `TypeError`.
* `ConvertPipe` uses these conversion semantics with Pipes.
* Add `PacketList.convert_to` (which calls `Packet.convert_to`) and `PacketList.getlayer` (which calls `Packet.getlayer`).

All of these conversions are **not** guaranteed to be lossless, nor are they guaranteed to be bidirectional.  It is intended to convert between two functionally similar packet types.

This is used by the Nordic BLE protocol handling branch I'm working on (nrf5sniff), to convert between the protocol used by the device and the protocol declared by `DLT_NORDIC_BLE` (and a couple of other variants). As that's a much larger patch, I've sharded it up to make it easier to review.

Adds unit tests for all of this, which also serve as examples as to how this could be used.

p-l- commented:
Hi,

Thanks for submitting this PR. What are some use cases for that?

micolous commented:
> What are some use cases for that?

I have more detailed examples in here: https://github.com/micolous/scapy/blob/nrf5sniff/scapy/contrib/nrf_sniffer.py

This is essentially a requirement for using that hardware -- Nordic has implemented the packet format differently between the hardware and `DLT_NORDIC_BLE`.

An example of using those conversions with Pipes:

```python3
load_contrib('nrf_sniffer')
s = nrfs_connect('/dev/ttyUSB0')  # this feature subject to patch, currently in nrf5sniff
p = PipeEngine()
src = SniffSource(socket=s)  # this feature is subject to another patch, currently in nrf5sniff
w = WiresharkSink()  # this sink subject to another patch, currently in nrf5sniff
c = ConvertPipe(NRFS2_PCAP)  # in this PR
p.add(src > c > w)
p.start()
```

gpotter2 commented:
It makes sense.

A note however: we already have `decode_payload_as` which does pretty much the same thing, but for the payload.
For consistency, it would be great if you could rename your function to `decode_as`

(Also I’m glad some people use the Pipes :smile: their documentation is a bit poor, and it doesn’t seem many people have ever posted about it :/)

micolous commented:
> A note however: we already have `decode_payload_as` which does pretty much the same thing, but for the payload.
> For consistency, it would be great if you could rename your function to `decode_as`

`decode_payload_as` is a little different to `convert_to`:

* `decode_payload_as` serialises the `payload` in binary form, and then sends it unmodified to another `Packet` subclass.

* `convert_to` implementations will typically read the fields of a packet type `A`, and map it onto equivalents in packet type `B`.

The `convert_to` for `Raw` would be equivalent to `decode_as`, but it is an exceptional case.  An ideal use-case would `scapy.contrib.opc_da`, where there are multiple implementations of `Packet`s based on endianness.

Typical `convert_to` usages may be lossy, for example:

* packet of type `A` has a field stored as a fixed-precision integer, but packet of type `B` uses IEEE-754 floating points for the field

* packet of type `A` has a field with no equivalent in packet type `B`

gpotter2 commented:
So from what I understand, It has no real usage by default, as you need to implement it manually to your packets.

That’s pretty much doable without adding a global class function to all packets. You could make your own function to convert packets. It raises so many questions of what it does behind the scene, that I would describe it as “undefined behavior”

I don’t think we want to include that into Scapy by default :/

micolous commented:
> So from what I understand, It has no real usage by default, as you need to implement it manually to your packets. That’s pretty much doable without adding a global class function to all packets. You could make your own function to convert packets.

"No real usage" right now (aside from tests).  I split this from the rest of the nRF code in order to get feedback on the API. :)

I could just implement this in nRF (and I did, originally), but I thought this would be a useful pattern that's worth moving further up the stack.  The nRF stuff isn't the only sniffer firmware around, and they may encapsulate `DLT_BLUETOOTH_LE_LL` in different ways.

> It raises so many questions of what it does behind the scenes, that I would describe it as “undefined behavior”. 

The only conversions implemented "by default" is to `Raw` to lower layers.  I didn't implement any automated field matching, because that would be _totally_ undefined behaviour.

The idea is that `Packet` subclasses _could_ explicitly implement it, if it made sense, such as between different versions or variants of a protocol.

Putting it in `Packet` is needed to build the link between `Packet.convert_to` and `Packet.(classmethod)convert_packet`.  This allows any Scapy module (including contribs) to register conversions for anything else, without further modifications or imports required of those modules.

guedou commented:
I think that it makes sense to include it in Scapy, even though there is only one use case so far.

micolous commented:
I've picked this up again today, and rebased on `master`.

Figured out the Windows test failure: `QueueSink` puts the packets from `high` and `low` into a single queue, [which lead to a race](https://ci.appveyor.com/project/secdev/scapy/builds/25471152/job/67vucdwbj7v7ht4m#L901).  [I have now fixed the race](https://ci.appveyor.com/project/secdev/scapy/builds/25471335).

There's a couple of unrelated [flakey](https://travis-ci.com/secdev/scapy/jobs/210215731#L3030) [tests](https://ci.appveyor.com/project/secdev/scapy/builds/25471412/job/b0p5qm1ksc2vl457#L836) -- I'm going to do another force-push to see if that shakes it out.

micolous commented:
Tests passing, please take a look @guedou

guedou commented:
It looks good to me. However, I have one question, what would be the result of converting a IntField to a ByteField ? I am afraid that this corner case won't be detected before calling raw().

gpotter2 commented on scapy/packet.py:
```diff
@@ -1407,6 +1407,56 @@ def command(self):
+            return Raw(raw(self))
+
+        if self.haslayer(other_cls):
+            return self.getlayer(other_cls)
```
I'm still a bit skeptical.
This behavior is strange to me :/

micolous replied:
That was fallback behaviour to deal with a couple of problems I had:

1. HDLC / PPP has a complex inheritance hierarchy.  I attempted to implement a PPP client in Scapy at one point, which included support for swtiching between using `HDLC`, `PPP`, and `PPP_` modes.

  I think that's more of a special case, so I'm going to leave that out.

2. One of the examples I wrote up for nRF was to extract `BTLE` packets from a `PacketList`.  I forgot to add the `PacketList.convert_to` method to this PR (which I've since added).

  I'll work around this by adding `PacketList.getlayer`, which will allow one to explicitly fetch some child layers.

I've now removed that.

gpotter2 commented on scapy/packet.py:
```diff
@@ -1407,6 +1407,56 @@ def command(self):
+
+        if "_internal" not in kwargs:
+            return other_cls.convert_packet(self, _internal=True, **kwargs)
+
```
Also I'd rather raise an error if the conversion isn't implemented

micolous replied:
Done.

gpotter2 commented:
Great job with the Automaton failure ! :smile:

micolous commented:
> It looks good to me. However, I have one question, what would be the result of converting a IntField to a ByteField ? I am afraid that this corner case won't be detected before calling raw().

I don't think that issue is applicable to `Raw`, it is the same as calling `Raw(raw(packet))`. Something like this throws an error, as it did before:

```python
packet = TCP(sport=99999)  # > 65535, invalid

# all of these will throw an error
raw(packet)
packet.convert_to(Raw)
Raw(raw(packet))
```

Conversion may be a lossy process, [as I noted this in my earlier comment](https://github.com/secdev/scapy/pull/1913#issuecomment-475934179).

An example of where this _will be lossy_ for [nRF Sniffer packets](https://github.com/micolous/scapy/blob/nrf5sniff/scapy/contrib/nrf_sniffer.py):

* Nordic's (Python) tools add an extra field to packets which corresponds to the `COM` port the device was detected on (even on non-Windows!), which the device itself does not emit.
* Conversion between `DLT_BLUETOOTH_LE_LL_WITH_PHDR` (`class BTLE_RF`) and nRF's equivalents are lossy in some metadata, as the data granularity differs.

That said, most of the packet conversion is lossless (rearranging fields).  I'd like it if nRF Sniffer did things differently, but changing this is now hard because it's baked into tools like Wireshark. :(
